File: infer/tests/codetoanalyze/cpp/bufferoverrun/external.cpp
/*
 * Copyright (c) 2017 - present Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */

extern char** lib();

void extern_bad() {
  int arr[10];
  char** p = lib();
  if (p != 0)
    arr[10] = 0;
  char* q = *p;
  if (q != 0)
    arr[20] = 0;
  int r = *q;
  if (r != 0)
    arr[30] = 0;
}

File: infer/tests/codetoanalyze/java/bufferoverrun/External.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import external.library.SomeExternalClass;

public class External {
  /* This function should raise deduplicated issues because the symbolic value of external method
   * should be instantiated to top. */
  void external_function_Bad(SomeExternalClass v) {
    int i = (int) (v.externalMethod1()) + 1 + (int) (v.externalMethod1()) + 1;
  }

  /* This function should have no proof obilgation. */
  void call_external_function_Good(SomeExternalClass v) {
    external_function_Bad(v);
  }
}
